[swift2objc] Const Foundation #2551 - fix

DPrakashhh opened the pull request:
I finally finished the Foundation parsing speedup we talked about in #2551. CI was taking way too long (minutes!), so I’ve focused on getting that run time down so we aren't just sitting around waiting for tests.

The big win: Total generation time is now around `2.27 seconds` on my latest run, down from several minutes in CI. The actual JSON load happens in just `0.89s`.

Here’s the breakdown of what I did:

**JSON Caching**: I’ve checked in the `Foundation.symbols.json` files directly. I spent some time testing gzip compression to keep the repo size down, but the decompression overhead actually made it slower than just reading the raw file, so I stuck with the uncompressed JSON for pure speed.

**Pathing Fixes**: I ran into a "gotcha" where `Platform.script` wasn't pointing to the right place during integration tests (it was hitting temp folders). I’ve updated the logic to be more robust, searching the package root and working directory so the cache is found reliably regardless of how the tool is invoked.

**Binary Serialization (The experiment)**: I wrote a binary serialization module to try and push the load time even lower. It’s functional, but I hit a snag where the JSON object structure gets a bit messy during reconstruction. Since the JSON cache is already hitting our performance goals and is completely stable, I’ve set the tool to gracefully fallback to JSON if the binary path fails. This way, we get the speedup now without risking stability.

**Instrumentation**: I added a `PerfTimer` utility that gives a nice hierarchical breakdown in the logs. It helped me track down the bottlenecks and will be useful for us to keep an eye on performance as the AST grows.

I also went through and cleaned up all the analyzer warnings (mostly long lines and doc comments), so `dart analyze` is completely green now. `All 75 tests are passing.`

**Let me know if you want me to keep digging into that binary reconstruction issue, or if you're happy with the 3-second JSON path for now!**

<img width="1122" height="190" alt="Screenshot 2025-12-31 at 11 39 43 AM" src="https://github.com/user-attachments/assets/4848b9c4-7c89-48d9-ac11-99b43393b23e" />

liamappelbe commented:
> The big win: Total generation time is now around `2.27 seconds` on my latest run, down from several minutes in CI. The actual JSON load happens in just `0.89s`.

I think you might be measuring different things. The speedup is significant, but not minutes to seconds. Out of curiosity, can you push a branch that includes the `PerfTimer` stuff, but not the caching changes? That way we can get an apples-to-apples comparison.

> Let me know if you want me to keep digging into that binary reconstruction issue, or if you're happy with the 3-second JSON path for now!

The JSON path is plenty fast enough. You can remove the binary serialization stuff.

Also, remove the `PerfTimer` stuff from this PR, since we already have the benchmark results and don't need that stuff in everyday operation. The logs are [pretty spammy](https://github.com/dart-lang/native/actions/runs/20613365447/job/59291381447?pr=2918).

The other thing this PR will need is adding a step in `.github/workflows/swift2objc.yaml` to verify that the checked-in JSON is up to date. That way we'll get an alert if we need to regenerate it.

DPrakashhh commented:
Thank you for the encouragement! I've decided to stick with it. I'm currently simplifying the implementation to keep it 'JSON-only' and resolving the integration test failures locally. I won't push again until I have a clean, passing suite that respects the simplicity you're looking for. Thanks for the guidance! - will try my best to solve this today

DPrakashhh commented:
@liamappelbe sir, I’ve just pushed the finalized version of this PR.

Key Updates:
1- Fixed Nested Type Assertions: I discovered the failures were due to missing relationship metadata. I’ve updated the generator to cache all 10 Foundation extension symbolgraphs (like `@Swift` and `@CoreFoundation`) and the loader to merge them. This resolved the assertions while keeping the transformer logic simple.
2- 'Human-Clean' Implementation: I’ve stripped out all experimental binary serialization and custom instrumentation. The solution is now strictly JSON-based as requested.
3- Robust Pathing: I implemented a simple package-root resolver for the cache that works reliably across `dart run` and `dart test`.
4- CI Verification: Added a workflow step to ensure the checked-in JSON stays synchronized with the tool.

All 75 tests are passing . Ready for your final review!

DPrakashhh commented on pkgs/swift2objc/tool/generate_foundation_ast.dart:
```diff
@@ -0,0 +1,206 @@
+      help: 'Verify cached symbolgraph matches freshly extracted version',
+      negatable: false,
+    )
+    ..addOption(
```
done

liamappelbe commented:
> @liamappelbe , I am getting a test error , should i need to make more changes?

Yeah, it looks like your generate script is producing different results when run on the github bot vs run on your local machine.

Here's how I debugged it, so you can do this yourself and take a look at the results. The full json is in the logs, so I downloaded the raw logs, pasted those lines into seperate files, ran them through a JSON formatter, and had a look. I tried putting them in a gist to share, but github is choking because they're too big.

Anyway, the first difference I noticed was they start with `metadata` and `module` sections, which are different between the machines:

Checked in JSON:
```json
"metadata":
{
    "formatVersion":
    {
        "major": 0,
        "minor": 6,
        "patch": 0
    },
    "generator": "Apple Swift version 6.2.1 (swiftlang-6.2.1.4.8 clang-1700.4.4.1)"
},
"module":
{
    "name": "Foundation",
    "platform":
    {
        "architecture": "arm64e",
        "vendor": "apple",
        "operatingSystem":
        {
            "name": "macosx",
            "minimumVersion":
            {
                "major": 26,
                "minor": 1
            }
        }
    }
},
```

Bot's generated JSON:
```json
"metadata":
{
    "formatVersion":
    {
        "major": 0,
        "minor": 6,
        "patch": 0
    },
    "generator": "Apple Swift version 6.1.2 (swiftlang-6.1.2.1.2 clang-1700.0.13.5)"
},
"module":
{
    "name": "Foundation",
    "platform":
    {
        "architecture": "aarch64",
        "vendor": "apple",
        "operatingSystem":
        {
            "name": "macosx",
            "minimumVersion":
            {
                "major": 15,
                "minor": 5
            }
        }
    }
},
```

One of the differences is that the swift compiler version is different. Another difference is the macOS version. Finally, the CPU architecture is different. These sections don't matter for swift2objc, so we could just remove them from the generated JSON, as long as the rest of the JSON was the same.

The bigger issue is that the rest of the JSON isn't the same at all. I thought one issue might be that the JSON was generated in a random order, so maybe we'd have to sort the symbols by their ID. But it's worse than that. After formatting, your checked in JSON is 3.69M lines long, and the bot's generated JSON is only 3.46M lines long. So it's generated a very different symbolgraph.

The differences are probably related to the differences in the compiler version, OS version, target architecture. If it was just the target architecture that caused the differences, then it might be practical to check in a symbolgraph for the handful of different architectures we support. But the compiler and OS versions make this impractical.

Given the modest time savings we're getting, I think we just call this a failed experiment and move on to something else.

DPrakashhh commented:
sure! Sorry for wasting too much time on this. Although i learned many things :)
